src/Magnum/Octree/Octree.hpp

```diff
@@ -0,0 +1,344 @@
+
+    _center = {0, 0, 0};
+    _radius = 0.f;
+    _depth = maxDepth;
+    _isLeafNode = true;
+
+    Vector3 min {points.at(0).min()};
+    Vector3 max {points.at(0).max()};
+
+    /* find minimal and maximal x,y,z in the data points */
+    for(auto point : points) {
+        min = Math::min(point.min(), min);
+        max = Math::max(point.max(), max);
+    }
+
+    const Vector3 radius = max - min;
+
+    /* bounds should be squared, not rectangular. Take the largest radius. */
+    _radius = radius.max() * 0.5f;
+
+    /* find the center of the bounding box */
+    _center = min + Vector3 {_radius};
+
+    return *this;
+}
+
+/* insert operation. At the moment every leaf has exactly one data point. Depth is unlimited. */
+template <typename T>
+Octree<T>& Octree<T>::insert(const Shapes::AxisAlignedBox3D& point, T data) {
+
+    /*
+     * Use a simple code for making traversion simple, by using bitwise or operations
+     * < means, that the childs coordinate is less than the coordinate of the center of the current node
+     * > means greater than
+     * e.g. child[3].x < center.x, child[3].y > center.y, child[3].z > center.z
+     * child:	0 1 2 3 4 5 6 7
+     * x:       < < < < > > > >
+     * y:       < < > > < < > >
+     * z:       < > < > < > < >
+     */
+
+    if(isLeafNode()) {
+        /* maximum depth is reached */
+        if(_depth == 0 || _entries.size() == 0) {
+            _entries.push_back(point);
+            _data.push_back(data);
+            return *this;
+        } else {
+            /* max depth isn't reached yet -> split node */
+
+            /*
+             * if node is a leaf, max depth isn't reached and it contains more than a point,
+             * it means that the contained points are too large to fit in a child node.
+             * This means, looking at only the first point, still works.
+             */
+
+            /* if max depth isn't reached, it means, in the leaf node was exactly one point */
+            const Shapes::AxisAlignedBox3D& oldPoint = _entries.at(0);
+            const T oldData = _data.at(0);
+
+            /*
+             * {-0.5, -0.5, -0.5},
+             * {-0.5, -0.5, +0.5},
+             * {-0.5, +0.5, -0.5},
+             * {-0.5, +0.5, +0.5},
+             * {+0.5, -0.5, -0.5},
+             * {+0.5, -0.5, +0.5},
+             * {+0.5, +0.5, -0.5},
+             * {+0.5, +0.5, +0.5}
+             */
+            Int octContainingBox = getOctantContainingBox(point);
+            Int octContainingOldBox = getOctantContainingBox(oldPoint);
+
+            if((octContainingOldBox != -1) || (octContainingBox != -1)) {
+
+                /* create the 8 new children */
+                for(int i = 0; i < 8; ++i) {
+                    _children[i] = new Octree{
+                        _center + (_radius * Vector3 {
+                            (i & 4 ? 0.5f : -0.5f),
+                            (i & 2 ? 0.5f : -0.5f),
+                            (i & 1 ? 0.5f : -0.5f)}),
```

Comment:
With mosra/magnum@247d8dba6f9fcbaf313e695d516c082ac70fb410 it might be possible to rewrite this as:

```cpp
Math::lerp(Vector3{-0.5f}, Vector3{0.5f}, Math::BoolVector<3>{UnsignedByte{i}})
```

... but it's in reverse order, so then something like `Math::swizzle<'z', 'y', 'x'>()` (or maybe there's some other more clever solution that avoids the reordering, I don't know). Mentioning it just in case you find it useful somewhere.